streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -389,6 +391,18 @@ public void close() {
         metrics.removeAllSensors();
     }
 
+    void closeProducer() {
+        if (exactlyOnceEnabled) {
+            try {
+                recordCollector.close();
+            } catch (final Throwable e) {
+                log.error("{} Failed to close producer: ", logPrefix, e);
+            }
+        } else {
+            throw new IllegalStateException("Tasks should only close producers if exactly-once semantics is enabled.");
```

Comment:
It's a check for bugs -- it should never happen that `closeProducer` get's called. I can also make it an assertion is you like this better.